Support multiple namespace

rkoyama1623 opened the pull request:
Support multiple namespace in `.dae` file.

# Background
Hi, all. I have `example.dae` file whose collada element is as follows:

```collada
<?xml version="1.0" encoding="utf-8"?>
<COLLADA xmlns="http://www.collada.org/2005/11/COLLADASchema" version="1.4.1"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <asset>
    <contributor>
      <author>Blender User</author>
      <authoring_tool>Blender 2.90.0 commit date:2020-08-31, commit time:11:26, hash:0330d1af29c0</authoring_tool>
    </contributor>
...
  <library_visual_scenes>
    <visual_scene id="Scene" name="Scene">
    ...
    </visual_scene>
  </library_visual_scenes>
  <scene>
    <instance_visual_scene url="#Scene"/>
  </scene>
</COLLADA>

```
Please note that this file has two name spaces, and this is not invalid since collada file is a kind of XML file and follows both COLLADA Schema and XML Schema.
However, when this collada file is parsed with PyCollada, `scene` attribute is `None`, and that causes an error in some software.
```python
inport collada
with open('example.dae') as f:
    c = collada.Collada(f)
assert c.scene is not None # raise AssertionError
```

# Cause of the empty `scene` attribute
PyCollada get collada namespace in these lines.
https://github.com/pycollada/pycollada/blob/master/collada/__init__.py#L222-L223
These lines select the first namespace as collada namespace, but that is not always correct.
The bug above was caused by the bug in which the XML Schema URI was used as collada schema.

# What this PR do?
In this PR, select URIs which include "COLLADA", sort them in descending order, and use the first URI as collada schema.
If there are multiple URIs including "COLLADA", the last one in alphabetical order is used.
This process is intended to select latest collada xsd as collada schema, since the file is assumed to support the latest collada version (I don't know whether multiple collada URI could be appear in one `.dae` file, but it might be possible).
For example, if one file has collada 1.4 schema and collada 1.5 schema, schema of collada 1.5 will be selected.
```
Collada 1.4: https://www.collada.org/2005/11/COLLADASchema/
Collada 1.5: https://www.collada.org/2008/03/COLLADASchema/ # This URI will be used as collada schema
```

Thank you

jterrace commented on collada/__init__.py:
```diff
@@ -219,8 +219,11 @@ def __init__(self,
+                namespace = next(iter(collada_namespaces))
             elif hasattr(xml_root, 'tag'):
                 # for xml.etree we need to extract ns from root tag
                 namespace = xml_root.tag.split('}')[0].lstrip('{')
```
what about when lxml is unavailable, for etree?

jterrace commented on collada/__init__.py:
```diff
@@ -219,8 +219,11 @@ def __init__(self,
-                namespace = next(iter(xml_root.nsmap.values()))
+                # use the latest collada xsd in the namespace map
+                namespaces = xml_root.nsmap.values()
+                collada_namespaces = filter(lambda uri: 'COLLADA' in uri.upper(), namespaces)
```
what if this filters down to nothing? would it be better to fall back to using the first one if none match "collada" string?

jterrace reviewed (changes requested):
Can you add some tests for this?

rkoyama1623 commented:
Thank you for your quick reply and review.
1. I added test code for .dae file with multiple namespace (14d358f119f9ffa87cc47db0bad68f1a5b27d9d3).
By this commit, test_collada.py fails when lxml is used as xml parser.
2. I change the way to get namespace. (e2a48b864af9b9853dfa955fd1e8c31c1e65b7f3)
In this commit, we stop using `nsmap` of lxml to support multiple namespaces and the namespace is now extracted from the root tag (assumed to be COLLADA) in the same way whether lxml or xml is used.

Files without namespace declarations will generate an error, but that has been the case in the past. 
So we do not support it at this time.
I have confirmed that the test passes.

```bash
cd pycollada/collada/tests
for f in *.py; do python $f; done
```

jterrace commented on collada/__init__.py:
```diff
@@ -215,15 +215,12 @@ def __init__(self,
-            elif hasattr(xml_root, 'tag'):
-                # for xml.etree we need to extract ns from root tag
-                namespace = xml_root.tag.split('}')[0].lstrip('{')
+            root_tag = xml_root.tag
```
looks like this variable is unused?

rkoyama1623 commented:
Thank you for your review. I revised according to your advice.

jterrace commented:
thanks!
